homeassistant/components/modbus/sensor.py

```diff
@@ -121,86 +121,79 @@ async def async_setup_platform(
             del entry[CONF_REGISTER_TYPE]
 
     for entry in discovery_info[CONF_SENSORS]:
-        if entry[CONF_DATA_TYPE] == DATA_TYPE_STRING:
-            structure = str(entry[CONF_COUNT] * 2) + "s"
-        elif entry[CONF_DATA_TYPE] != DATA_TYPE_CUSTOM:
-            try:
-                structure = f">{DEFAULT_STRUCT_FORMAT[entry[CONF_DATA_TYPE]][entry[CONF_COUNT]]}"
-            except KeyError:
-                _LOGGER.error(
-                    "Unable to detect data type for %s sensor, try a custom type",
-                    entry[CONF_NAME],
-                )
-                continue
-        else:
-            structure = entry.get(CONF_STRUCTURE)
-
-        try:
-            size = struct.calcsize(structure)
-        except struct.error as err:
-            _LOGGER.error("Error in sensor %s structure: %s", entry[CONF_NAME], err)
-            continue
-
-        if entry[CONF_COUNT] * 2 != size:
-            _LOGGER.error(
-                "Structure size (%d bytes) mismatch registers count (%d words)",
-                size,
-                entry[CONF_COUNT],
-            )
-            continue
-
-        if CONF_REVERSE_ORDER in entry:
-            if entry[CONF_REVERSE_ORDER]:
-                entry[CONF_SWAP] = CONF_SWAP_WORD
-            else:
-                entry[CONF_SWAP] = CONF_SWAP_NONE
-            del entry[CONF_REVERSE_ORDER]
-        if entry.get(CONF_SWAP) != CONF_SWAP_NONE:
-            if entry[CONF_SWAP] == CONF_SWAP_BYTE:
-                regs_needed = 1
-            else:  # CONF_SWAP_WORD_BYTE, CONF_SWAP_WORD
-                regs_needed = 2
-            if (
-                entry[CONF_COUNT] < regs_needed
-                or (entry[CONF_COUNT] % regs_needed) != 0
-            ):
-                _LOGGER.error(
-                    "Error in sensor %s swap(%s) not possible due to count: %d",
-                    entry[CONF_NAME],
-                    entry[CONF_SWAP],
-                    entry[CONF_COUNT],
-                )
-                continue
         if CONF_HUB in entry:
             # from old config!
             hub = hass.data[MODBUS_DOMAIN][entry[CONF_HUB]]
         else:
             hub = hass.data[MODBUS_DOMAIN][discovery_info[CONF_NAME]]
         if CONF_SCAN_INTERVAL not in entry:
             entry[CONF_SCAN_INTERVAL] = DEFAULT_SCAN_INTERVAL
-        sensors.append(
-            ModbusRegisterSensor(
-                hub,
-                entry,
-                structure,
-            )
-        )
-
+        try:
+            sensors.append(ModbusRegisterSensor(hub, entry))
+        except ValueError as error:
+            _LOGGER.error(str(error), exc_info=True)
     if not sensors:
         return
     async_add_entities(sensors)
 
 
```

Comment:
This should be part of the class not outside.

Reply:
The idea was to skip the bad sensor configuration and continue with the other sensors. Basically, keep the behavior the same as before.

What would be the preference here:
- throw an exception in the __init__ method of the ModbusRegisterSensor and fail the sensor initialization. More like all or nothing approach but it will change the current logic
- allow creating the sensor but always return `available` False. This will change the behavior as well but in a good way. It won't be a situation when the sensor entity comes and go because of the configuration error.
- add __new__ method. Return None if the configuration is bad and filter out the sensor from the array later. This will be the closest approach but will make the sensor code harder to understand because of both __new__ and __init__ methods in the class.

Reply:
I prefer the second, with a _LOGGER.warning.